src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionHelper.cs

```diff
@@ -141,108 +82,155 @@ private static string GetReceiverTypeName(ITypeSymbol typeSymbol)
             {
-                // We use a very simple approach to build the cache in the background:
-                // queue a new task only if the previous task is completed, regardless of what
-                // that task is.
                 lock (s_gate)
                 {
+                    // We use a very simple approach to build the cache in the background:
+                    // queue a new task only if the previous task is completed, regardless of what
+                    // that task is.
                     if (s_indexingTask.IsCompleted)
                     {
-                        s_indexingTask = Task.Run(() => TryGetIndicesAsync(document.Project, forceIndexCreation: true, CancellationToken.None));
+                        s_indexingTask = PopulateIndicesAsync(document.Project, CancellationToken.None);
                     }
                 }
-
-                return (ImmutableArray<SerializableImportCompletionItem>.Empty, counter);
             }
 
-            counter.GetFilterTicks = Environment.TickCount - ticks;
-            counter.NoFilter = !indicesResult.HasResult;
+            return (itemsBuilder.ToImmutable(), counter);
 
-            ticks = Environment.TickCount;
+            static SerializableImportCompletionItem CreateItem(IMethodSymbol methodSymbol, Dictionary<INamespaceSymbol, string> stringCache, CancellationToken cancellationToken)
+                => new SerializableImportCompletionItem(
+                    SymbolKey.CreateString(methodSymbol, cancellationToken),
+                    methodSymbol.Name,
+                    methodSymbol.Arity,
+                    methodSymbol.GetGlyph(),
+                    GetFullyQualifiedNamespaceName(methodSymbol.ContainingNamespace, stringCache));
 
-            var items = await GetExtensionMethodItemsAsync(
-                document, receiverTypeSymbol, GetReceiverTypeNames(receiverTypeSymbol), indicesResult, position,
-                namespaceInScope, counter, cancellationToken).ConfigureAwait(false);
+            // Force create all relevant indices
+            static async Task PopulateIndicesAsync(Project currentProject, CancellationToken cancellationToken)
+            {
+                var solution = currentProject.Solution;
+                var cacheService = GetCacheService(solution.Workspace);
 
-            counter.GetSymbolTicks = Environment.TickCount - ticks;
+                foreach (var project in GetAllRelevantProjects(currentProject))
+                {
+                    _ = await GetCacheEntryAsync(project, loadOnly: false, cacheService, cancellationToken).ConfigureAwait(false);
+                }
 
-            return (items, counter);
+                foreach (var peReference in GetAllRelevantPeReferences(currentProject))
+                {
+                    _ = await SymbolTreeInfo.GetInfoForMetadataReferenceAsync(solution, peReference, loadOnly: false, cancellationToken).ConfigureAwait(false);
```

Comment:
have you considered doing these bits in parllel as well?

Reply:
That makes sense. Will do